src/paths/shortest_paths.c

```diff
@@ -1018,20 +1024,31 @@ int igraph_diameter(const igraph_t *graph, igraph_real_t *pres,
     if (pto != 0) {
         *pto = (igraph_integer_t) to;
     }
-    if (path != 0) {
+    if ((vertex_path) || (edge_path)) {
         if (! igraph_finite(res)) {
-            igraph_vector_clear(path);
+            if (vertex_path) {
+                igraph_vector_clear(vertex_path);
+            }
+            if (edge_path){
+                igraph_vector_clear(edge_path);
+            }
         } else {
-            igraph_vector_ptr_t tmpptr;
-            igraph_vector_ptr_init(&tmpptr, 1);
-            IGRAPH_FINALLY(igraph_vector_ptr_destroy, &tmpptr);
-            VECTOR(tmpptr)[0] = path;
-            IGRAPH_CHECK(igraph_get_shortest_paths(graph, &tmpptr, 0,
-                                                   (igraph_integer_t) from,
-                                                   igraph_vss_1((igraph_integer_t)to),
-                                                   dirmode, 0, 0));
-            igraph_vector_ptr_destroy(&tmpptr);
-            IGRAPH_FINALLY_CLEAN(1);
+            if ((vertex_path != 0) && (edge_path != 0)) {
+                IGRAPH_CHECK(igraph_get_shortest_path(graph, vertex_path, edge_path,
+                                                    (igraph_integer_t) from,
+                                                    (igraph_integer_t)to,
+                                                    dirmode));
+            } else if (vertex_path != 0) {
+                IGRAPH_CHECK(igraph_get_shortest_path(graph, vertex_path, 0,
+                                                    (igraph_integer_t) from,
+                                                    (igraph_integer_t)to,
+                                                    dirmode));
+            } else if (edge_path != 0) {
+                IGRAPH_CHECK(igraph_get_shortest_path(graph, 0, edge_path,
+                                                    (igraph_integer_t) from,
+                                                    (igraph_integer_t)to,
+                                                    dirmode));
+            }
```

Comment:
It seems to me that we don't need if `if ... else` construct here. Just call `igraph_get_shortest_path` directly with `vertex_path` and `edge_path`.